py3status/modules/tor_rate.py

```diff
@@ -0,0 +1,132 @@
+# -*- coding: utf-8 -*-
+"""
+Display the current transfer rates of a tor instance
+
+Configuration parameters:
+    cache_timeout: An integer specifying the cache life-time of the modules
+        output in seconds (default 2)
+    control_address: The address on which the Tor daemon listens for control
+        connections (default "127.0.0.1")
+    control_password: The password to use for the Tor control connection
+        (default None)
+    control_port: The port on which the Tor daemon listens for control
+        connections (default 9051)
+    format: A string describing the output format for the module
+        (default "↑ {up} ↓ {down}")
+    rate_unit: The unit to use for the transfer rates
+        (default "B/s")
+    si_units: A boolean value selecting whether or not to use SI units
+        (default 0)
+    value_format: A string describing how to format the transfer rates
+        (default "[\?min_length=12 {rate:.1f} {unit}]")
+
+Format placeholders:
+    {down} The incoming transfer rate
+    {up} The outgoing transfer rate
+
+Requires:
+    stem: python module from pypi https://pypi.python.org/pypi/stem
+
+Example:
+
+```
+tor_rate {
+    cache_timeout = 10
+    format = "IN: {down} | OUT: {up}"
+    control_port = 1337
+    control_password = "TertiaryAdjunctOfUnimatrix01"
+    si_units = 1
+}
+
+order += "tor_rate"
+```
+
+@author Felix Morgner <felix.morgner@gmail.com>
+@license 3-clause-BSD
+"""
+
+from stem import ProtocolError, SocketError
+from stem.connection import AuthenticationFailure
+from stem.control import Controller, EventType
+
+
+class Py3status:
+    cache_timeout = 2
+    control_address = '127.0.0.1'
+    control_password = None
+    control_port = 9051
+    format = u'↑ {up} ↓ {down}'
+    rate_unit = 'B/s'
+    si_units = False
+    value_format = '[\?min_length=12 {rate:.1f} {unit}]'
+
+    def __init__(self):
+        self._down = 0
+        self._handler_active = False
+        self._up = 0
+
+    def tor_rate(self, outputs, config):
+        text = ''
+        if not self._handler_active:
+            try:
+                self._register_event_handler()
```

Comment:
is this something we want to keep doing if it fails?  for some it feels ok eg no connection, but for authentication fail should we just give up?